ezrassor:  
  controller_manager:
    ros__parameters:
      update_rate: 1000  # Hz

      diff_drive_controller:
        type: diff_drive_controller/DiffDriveController

      joint_state_broadcaster:
        type: joint_state_broadcaster/JointStateBroadcaster

      arm_back_velocity_controller:
        type: velocity_controllers/JointGroupVelocityController

      drum_back_velocity_controller:
        type: velocity_controllers/JointGroupVelocityController
      
      paver_arm_controller:
        type: velocity_controllers/JointTrajectoryController
      
      paver_arm_claw_controller:
        type: effort_controllers/JointGroupEffortController

  diff_drive_controller:
    ros__parameters:
      left_wheel_names: [ "left_wheel_front_joint", "left_wheel_back_joint" ]
      right_wheel_names: [ "right_wheel_front_joint", "right_wheel_back_joint" ]

      # Separation between a left and right wheel in meters
      wheel_separation: 0.10

      wheel_radius: 0.18

      wheel_separation_multiplier: 1.0
      left_wheel_radius_multiplier: 1.0
      right_wheel_radius_multiplier: 1.0

      publish_rate: 50.0
      odom_frame_id: odom
      base_frame_id: base_link
      pose_covariance_diagonal : [0.001, 0.001, 0.001, 0.001, 0.001, 0.01]
      twist_covariance_diagonal: [0.001, 0.001, 0.001, 0.001, 0.001, 0.01]

      open_loop: true
      enable_odom_tf: true

      cmd_vel_timeout: 0.5
      use_stamped_vel: false

      # Preserve turning radius when limiting speed/acceleration/jerk
      preserve_turning_radius: true

      # Publish limited velocity
      publish_cmd: true

      # Publish wheel data
      publish_wheel_data: true

      # Velocity and acceleration limits
      linear.x.has_velocity_limits: true
      linear.x.has_acceleration_limits: true
      linear.x.has_jerk_limits: false
      linear.x.max_velocity: 10.0
      linear.x.min_velocity: -10.0
      linear.x.max_acceleration: 2.0
      linear.x.max_jerk: 0.0
      linear.x.min_jerk: 0.0

      angular.z.has_velocity_limits: true
      angular.z.has_acceleration_limits: true
      angular.z.has_jerk_limits: false
      angular.z.max_velocity: 10.0
      angular.z.min_velocity: -10.0
      angular.z.max_acceleration: 2.0
      angular.z.min_acceleration: -2.0
      angular.z.max_jerk: 0.0
      angular.z.min_jerk: 0.0
  
  arm_back_velocity_controller:
    ros__parameters:
        joints: 
          - arm_back_joint
        pid: {p: 100.0, i: 0.01, d: 10.0}
  
  drum_back_velocity_controller:
    ros__parameters:
        joints: 
          - drum_back_joint
        pid: {p: 100.0, i: 0.01, d: 10.0}

  # Paver arm controller using the joint trajectory controller
  paver_arm_controller:
    ros__parameters:
      joints:
      - joint12
      - joint23
      - joint34
      - joint45
      - joint56

      constraints:
        goal_time: 0.5
        stopped_velocity_tolerance: 0.02
        joint12:
          trajectory: 0.05
          goal: 0.02
        joint23:
          trajectory: 0.05
          goal: 0.02
        joint34:
          trajectory: 0.05
          goal: 0.02
        joint45:
          trajectory: 0.05
          goal: 0.02
        joint56:
          goal: 0.01
      gains:
        joint1: {p: 100, d: 1, i: 1, i_clamp: 1}
        joint2: {p: 100, d: 1, i: 1, i_clamp: 1}
        joint3: {p: 100, d: 1, i: 1, i_clamp: 1}
        joint4: {p: 100, d: 1, i: 1, i_clamp: 1}
        joint5: {p: 100, d: 1, i: 1, i_clamp: 1}
      stop_trajectory_duration: 0.5
      state_publish_rate: 25
      action_monitor_rate: 10
  
  paver_arm_claw_controller:
    ros__parameters:
      joints: 
        - grabber_joint1
        - grabber_joint2
      grabber_joint1: {p: 10.0, i: 0.00, d: 0.0}
      grabber_joint2: {p: 10.0, i: 0.00, d: 0.0}